### DISCUSSION TITLE
How to handle a helper and a modifier with the same name?

### DISCUSSION THREAD

**Discussion by boris-petrov** [typed-ember/glint#204] [Q&A] [answered]
I'm using both [@ember/render-modifiers](https://github.com/emberjs/ember-render-modifiers/) and [ember-render-helpers](https://github.com/buschtoens/ember-render-helpers). The former adds a `did-insert` modifier (and obviously `did-update` and `will-destroy`) and the latter - a `did-insert` helper.

I don't think it's possible to handle that in Glint right now. What can I do?

**Comment by dfreeman** [comment, marked as answer]
I think this is something we're probably not super likely to officially support, as it's not really clear how to do it in a clean way and this problem goes away with strict mode.

That said, _unofficially_ this is expressible using private APIs. As always, **use these at your own risk**—these types might change in any Glint release.

```ts
// ️⚠️ This code may break at any time! This version is current as of Glint 1.0.0-beta.1

import { ModifierReturn, DirectInvokable } from '@glint/template/-private/integration';
import { ContentValue } from '@glint/template';

declare module '@glint/environment-ember-loose/registry' {
  export default interface Registry {
    'did-insert': DirectInvokable<{
      (callback: () => void): ContentValue;
      <T extends Element>(element: T, callback: (el: T) => void): ModifierReturn;
    }>;
  }
}

```

**Reply by boris-petrov** [reply]
@dfreeman this doesn't work any more on Glint 0.9.1. :disappointed: There is at least one problem - `EmittableValue` doesn't exist in `@glint/template/-private/dsl`. Perhaps there are others. Do you mind updating the example to work with the latest Glint version?

**Reply by chriskrycho** [reply]
In the meantime, I will reiterate an alternative approach: *update your code*. This is likely to remain fragile and to break regularly, and while we will try to help folks stay unblocked, we also do not take it to be our responsibility to support every single pattern which exists in the wild—as Dan noted at the outset, this uses private API and broke in *exactly the way we said it could*.

In this case: the use of those modifiers and helpers are both *known anti-patterns in Octane*, and you should move away from them *anyway*.

**Reply by boris-petrov** [reply]
Actually I hit this with ember-keyboard’s `on-key` helper/modifier pair. In any case, if that’s not a supported use case, I’ll just suppress the errors for now. Thanks!

**Reply by dfreeman** [reply]
I think all you should need to change for this to keep working is to replace use of `EmittableValue` to `import { ContentValue } from '@glint/template'`, since we have a public type for that now. If you can confirm that that works, I'll update the snippet here.

**Reply by boris-petrov** [reply]
It seems to work, yes. For reference, here's how I implemented the `on-key` helper/modifier from `ember-keyboard`:

```ts
type OnKeyHelperAndModifier = DirectInvokable<
  (
    args: { readonly priority?: number },
    key: string,
    callback: (event: KeyboardEvent, emberKeyboardEvent: KeyboardEvent) => void,
  ) => BoundModifier<Element> & ContentValue
>;
```

Thanks!

**Reply by chriskrycho** [reply]
I learned via this discussion that we made this public API in the most recent release, so I stand corrected (and happily so)!

**Reply by dfreeman** [reply]
Just to be extra clear for any future travelers: `ContentValue` (and `AttrValue`) are public, but the rest of this is still very much internal and subject to change at any time! 😄

**Comment by ijlee2** [comment]
An update for those who use `@ember/render-modifiers` and `ember-render-helpers`: I renamed the helpers in `ember-render-helpers@1.0.0` so that there won't be name conflicts.

https://github.com/buschtoens/ember-render-helpers/releases/tag/v1.0.0
